src/typing/classy.ml

```diff
@@ -0,0 +1,252 @@
+         the getter. Otherwise just use the getter type or the setter type *)
+      let getters = SMap.map Func_sig.gettertype s.getters in
+      let setters = SMap.map Func_sig.settertype s.setters in
+      let getters_and_setters = SMap.fold (fun name t ts ->
+        match SMap.get name ts with
+        | Some t' -> Flow.unify cx t t'; ts
+        | None -> SMap.add name t ts
+      ) setters getters in
+
+      (* Treat getters and setters as fields *)
+      let fields = SMap.map fst s.fields in
+      let fields = SMap.union getters_and_setters fields in
+
+      fields, methods
+    )
+    in
+    let class_id = if static then 0 else s.id in
+    let constructor = if static then None else
+      let ts = List.rev_map Func_sig.methodtype s.constructor in
+      match ts with
+      | [] -> None
+      | [t] -> Some t
+      | t::_ as ts ->
+        let open Type in
+        let t = IntersectionT (reason_of_t t, InterRep.make ts) in
+        Some t
+    in
+    let fields, methods = elements ?constructor ~static cx s in
+    { Type.
+      class_id;
+      type_args = s.tparams_map;
+      arg_polarities = arg_polarities s;
+      fields_tmap = Flow.mk_propmap cx fields;
+      methods_tmap = Flow.mk_propmap cx methods;
+      mixins = false;
+      structural = Classy.structural;
+    }
+
+  let classtype cx x =
+    let x = Classy.remove_this x in
+    let { Sig.
+      tparams;
+      static = { Sig.reason = sreason; super = ssuper; _ };
+      instance = { Sig.reason; super; _ };
+      _;
+    } = x in
+    let open Type in
+    let sinsttype, insttype = Sig.mutually (insttype cx x) in
+    let static = InstanceT (sreason, MixedT.t, ssuper, sinsttype) in
+    let this = InstanceT (reason, static, super, insttype) in
+    (if Classy.ct_check_polarity then Flow.check_polarity cx Positive this);
+    let t = Classy.mk_class this in
+    if tparams = [] then t else PolyT (tparams, t)
+
+  let generate_tests cx f x =
+    let open Sig in
+    let {tparams; tparams_map; constructor; static; instance; _} = x in
+    Flow.generate_tests cx instance.reason tparams (fun map -> f {
+      x with
+      tparams_map = SMap.map (Flow.subst cx map) tparams_map;
+      constructor = List.map (Func_sig.subst cx map) constructor;
+      static = Classy.subst_sig cx map static;
+      instance = Classy.subst_sig cx map instance;
+    })
+  let check_super cx x =
+    let open Sig in
+    let x = Classy.remove_this x in
+    let reason = x.instance.reason in
+    mutually (fun ~static ->
+      let super = with_sig ~static (fun s -> s.super) x in
+      let insttype = insttype ~static cx x in
+      Flow.flow cx (super, Type.SuperT (reason, insttype))
+    ) |> ignore
+end
+
+module Class = Make(Class_sig.T)
+module Interface = Make(Iface_sig.T)
+module DeclClass = Make(Decl_class_sig.T)
+
+(* Processes the bodies of instance and static class members. *)
+let toplevels cx ~decls ~stmts ~expr x =
```

Comment:
I hacked this function to make it work. @samwgoldman had it so pretty--definitely needs tidying.